### PR TITLE
[#163591981] Allow users to search messages that contain a text

### PR DESCRIPTION
**Pull request by fpersico**
search

### CODE REVIEW

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#163591981](https://www.pivotaltracker.com/story/show/163591981): Come CIT voglio poter ricercare fra i messaggi

<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

**Comment by cloudify** [comment]
The search above the title looks kind of weird, @matteodesanti what do you think?

File: ts/components/messages/MessagesSearch.tsx [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+type State = {
+  lastMessagesState: ReturnType<typeof lexicallyOrderedMessagesStateSelector>;
+  lastSearchText: Option<string>;
+  filteredMessageStates: ReturnType<
+    typeof generateMessagesStateMatchingSearchTextArray
+  >;
+};
+
+/**
+ * Filter only the messages that matchs the searchText.
```
Comment (cloudify) [inline ts/components/messages/MessagesSearch.tsx]:
```suggestion
 * Filter only the messages that match the searchText.
```

File: ts/components/messages/MessagesSearch.tsx [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+ * The searchText is checked both in message and in service properties.
+ */
+const generateMessagesStateMatchingSearchTextArray = (
+  potMessagesState: pot.Pot<ReadonlyArray<MessageState>, string>,
+  servicesById: ServicesByIdState,
+  searchText: string
+): ReadonlyArray<MessageState> => {
+  return searchText === ""
```
Comment (cloudify) [inline ts/components/messages/MessagesSearch.tsx]: I think we should require at least a few characters for the search keyword, it doesn't really make sense to search for a keyword of less than 3 characters.

 ```suggestion
  return searchText.length < 3
```

File: ts/components/messages/MessagesSearch.tsx [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+              false
+            )
+          )
+        ),
+        []
+      );
+};
+
+/**
+ * A component to render a list of messages that matches a searchText.
```
Comment (cloudify) [inline ts/components/messages/MessagesSearch.tsx]:
```suggestion
 * A component to render a list of messages that match a searchText.
```

File: ts/screens/messages/MessagesHomeScreen.tsx [hunk 4]
```diff
@@ -39,12 +45,49 @@ const styles = StyleSheet.create({
...
+      <TopScreenComponent
+        title={I18n.t("messages.contentTitle")}
+        icon={require("../../../img/icons/message-icon.png")}
+        headerBody={
+          searchText.isSome() ? (
+            <Item>
+              <Input
+                placeholder={I18n.t("global.actions.search")}
+                value={searchText.value}
+                onChangeText={this.onSearchTextChange}
```
Comment (cloudify) [inline ts/screens/messages/MessagesHomeScreen.tsx]: we probably need to debounce this event, I'd say with a 500ms delay

File: ts/screens/messages/MessagesHomeScreen.tsx [hunk 5]
```diff
@@ -55,48 +98,87 @@ class MessagesHomeScreen extends React.Component<Props, never> {
...
+    const { searchText } = this.state;
+
+    return (
+      <MessagesSearch
+        messagesState={lexicallyOrderedMessagesState}
+        servicesById={servicesById}
+        paymentByRptId={paymentsByRptId}
+        onRefresh={refreshMessages}
+        navigateToMessageDetail={navigateToMessageDetail}
+        searchText={searchText.getOrElse("")}
```
Comment (cloudify) [inline ts/screens/messages/MessagesHomeScreen.tsx]: if `searchText` is empty, why render this component at all?

Comment (cloudify) [reply ts/screens/messages/MessagesHomeScreen.tsx]: also, can we have a loading indicator somewhere until the search results are ready?

**Comment by fpersico** [comment]
Updated version with:

- debounce on search bar input (500ms as suggested)
- loading indicator also when filtering
- search bar input must have at least three characters

search

File: ts/components/messages/MessagesSearch.tsx [hunk 6]
```diff
@@ -45,92 +44,116 @@ const generateMessagesStateMatchingSearchTextArray = (
...
+    const { messagesState, servicesById, searchText } = this.props;
+    const { potFilteredMessageStates } = this.state;
+
+    // Set filtering status
+    this.setState({
+      potFilteredMessageStates: pot.toLoading(potFilteredMessageStates)
+    });
+
+    // Start filtering messages
+    const filteredMessageStates = await Promise.resolve(
```
Comment (cloudify) [inline ts/components/messages/MessagesSearch.tsx]: what's the point of `await Promise.resolve` ?

**Pull request merged**